esp/services/ws_workunits/ws_workunitsService.cpp

```diff
@@ -4244,26 +4244,54 @@ inline StringBuffer &buildFullDllPath(StringBuffer &dllpath, StringBuffer &dllna
     return addPathSepChar(dllpath.set(dir)).append(sharedObjectFileName(dllname, name, ext, copy));
 }
 
-void writeSharedObject(const char *srcpath, const MemoryBuffer &obj, const char *dir, StringBuffer &dllpath, StringBuffer &dllname, unsigned crc)
+void writeTempSharedObject(const MemoryBuffer &obj, const char *dir, StringBuffer &filename)
+{
+    OwnedIFileIO io = createUniqueFile(dir, "query_copy_dll_", NULL, filename);
+    io->write(0, obj.length(), obj.toByteArray());
+}
+
+void writeSharedObject(const char *srcpath, const MemoryBuffer &obj, const char *dir, StringBuffer &dllpath, StringBuffer &dllname)
 {
     StringBuffer name, ext;
     if (srcpath && *srcpath)
         splitFilename(srcpath, NULL, NULL, &name, &ext);
 
     unsigned copy=0;
     buildFullDllPath(dllpath.clear(), dllname.clear(), dir, name.str(), ext.str(), copy);
-    while (checkFileExists(dllpath.str()))
+
+    unsigned crc=0;
+    StringBuffer tempDllName;
+    const unsigned attempts = 3; // max attempts
+    for (unsigned i=0; i<attempts; i++)
     {
-        if (crc && crc == crc_file(dllpath.str()))
+        while (checkFileExists(dllpath.str()))
+        {
+            if (crc==0)
+                crc = crc32(obj.toByteArray(), obj.length(), 0);
+            if (crc == crc_file(dllpath.str()))
+            {
+                DBGLOG("Workunit dll already exists: %s", dllpath.str());
+                if (tempDllName.length())
+                    removeFileTraceIfFail(tempDllName);
+                return;
+            }
+            buildFullDllPath(dllpath.clear(), dllname.clear(), dir, name.str(), ext.str(), ++copy);
+        }
+        try
         {
-            DBGLOG("Workunit dll already exists: %s", dllpath.str());
+            if (!tempDllName.length())
+                writeTempSharedObject(obj, dir, tempDllName);
```

Comment:
Unlikely:  This write should probably be outside the try block - otherwise on disk full the first write will fail, on the second time it will rename the empty file.